File: src/nahyeon/TwoPointers/boj_1806_부분합.java
package nahyeon.TwoPointers;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_1806_부분합 {

    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N, S;
    static int A[];

    static void input() {
        N = scan.nextInt();
        S = scan.nextInt();
        A = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            A[i] = scan.nextInt();
        }
    }

    static void pro() {
        int R = 0, sum = 0, ans = N + 1;
        for (int L = 1; L <= N; L++) {
            // L - 1 부분 빼기
            sum -= A[L-1];
            // R 을 조건에 만족할 때 까지 넘기기
            while (R + 1 < N && sum < S) {
                sum += A[++R];
            }
            // 조건이 만족하면 정답을 갱신한다.
            if (sum > S)
                ans = Math.min(ans, R - L + 1);
        }
        // 해당되는 조건이 없다면 0으로 갱신한다.
        if (ans == N + 1)
            ans = 0;
        System.out.println(ans);
    }

    public static void main(String[] args) {
        input();
        pro();
    }

    static class FastReader {

        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreTokens()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }
}

File: src/hyewon/TwoPointers/boj_1806_부분 합.kt
package hyewon.TwoPointers

/**
 * 주저리
 * L이 1일 때, R이 5에서 멈춘당. => L을 2로 땡겼을 때, R을 굳이 2부터 다시 시작할 필요가 있을까? => X => 5부터 시작하면 된다.
 */
fun main() = with(System.`in`.bufferedReader()) {
    val (N, S) = readLine().split(" ").map { it.toInt() }
    val A = ("0 " + readLine()).split(" ").map { it.toInt() }

    var R = 0
    var sum = 0
    var ans = N + 1
    for (L in 1..N) {
        // for문에 의해 L이 1증가했으므로, L - 1을 구간에서 제외
        sum -= A[L - 1]

        // R을 옮길 수 있을 때 까지 옮긴다.
        // R의 오른쪽 칸(R + 1)이 구간을 벗어나면 안되고, sum이 S이상이 될 때 까지
        while (R + 1 <= N && sum < S) {
            R++ // R을 옮겨주고
            sum += A[R] // sum에 새로 들어온 A[R]을 더함
        }

        // [L..R]의 합 즉, sum이 조건을 만족하면 정답 갱신
        if (sum >= S) ans = minOf(ans, R - L + 1)
    }

    if (ans == N + 1) ans = 0

    println(ans)
}